### PR TITLE
Mobius schedule

### PR DESCRIPTION
**Pull request by gajagajago**
## Description
This PR separates SpiralPipe with/without remapping. SpiralPipe without remapping resembles Mobius. Also, recomputation and separating optimizers (a.k.a spiral stage optimizer) is activated through argument. This PR also resolves some known issues below:
- Fix dropout resulting in non-identical result from recomputation compared to orginal fwd. 
- Fix recv wait in compute stream to guarantee correctness.

### CODE REVIEW

File: megatron/spiral/initialize.py [hunk 1]
```diff
@@ -172,7 +172,28 @@ def wait(self):
...
+            and self.record_stream_event_deque[0].event.query()
+        ):
+            completed_event_hdl = self.record_stream_event_deque.popleft()
+            if completed_event_hdl == self:
+                break
+
+        if self.post_wait_fn is not None:
+            self.post_wait_fn()
+
+    def synchronize(self):
```
Comment (ttochi) [inline megatron/spiral/initialize.py:187]: It seems duplicated with line 208

File: megatron/training.py [hunk 2]
```diff
@@ -638,9 +691,18 @@ def setup_model_and_optimizer(model_provider_func,
...
+    if args.spiral:
+        if args.spiral_stage_optimizer:
+            from megatron.spiral.optimizer import SpiralStageOptimizer
+
+            assert isinstance(
+                optimizer, SpiralStageOptimizer
+            ), f"Optimizer must be SpiralStageOptimizer for spiral_stage_optimizer, but got {type(optimizer)}"
+            assert (
+                len(optimizer.optimizer_list) == args.spiral_backward_virtual_size
+            ), f"Number mismatch for optimizers in SpiralStageOptimizer, requires {args.spiral_backward_virtual_size} but got {len(optimizer.optimizer_list)}"
```
Comment (ttochi) [inline megatron/training.py:705]: Wouldn't it better check assertion in `get_megatron_optimizer`?

File: megatron/spiral/schedules.py [hunk 3]
```diff
@@ -743,32 +794,517 @@ def cleanup():
...
+
+        # spiral stage optimizer
+        if optimize_after_bwd_stage:
+            torch.cuda.nvtx.range_push(f"opt b[{bwd_stage_id}]")
+            _optimizer_process = Process(
+                target=_weight_update,
+                args=(
+                    optimizer,
+                    bwd_stage_id,
+                    offload_event_queries.pop(0), # TODO (SpiralPipe) This is very error-prone, since here and prefetch stream both pops head from the offload event queries. Pop after querying with appropriate tag.
```
Comment (ttochi) [inline megatron/spiral/schedules.py:1292]: When I run mobius-recompute with --stage-optimizer, I get this error
```
IndexError: pop from empty list
```

It seems that we need to add "optimizer:b{bwd_stage_id}" event before push "offload:b{bwd_stage_id}" event (like in L743)

Comment (gajagajago) [reply megatron/spiral/schedules.py:1292]: oh sorry for mistake. i will fix it asap

File: megatron/training.py (hunk 2 above)
Comment (gajagajago) [reply megatron/training.py]: removed assertion since too trivial

**Comment by gajagajago** [comment]
@ttochi I have resolved all known issues (including the one illustrated on PR description) and merged your commits (on `experiment` branch) with some edits. Please check it out.

**Review by ttochi — APPROVED** [review]
LGTM
I commited event synchronize to ensure the optimizer.step run after last bwd stage's offload

**Pull request merged**